### PR TITLE
Add flag to close cmake window after successful build and call autocommand events

### PR DESCRIPTION
**Pull request by AlexAUT**
Implemented a flag to close the window after a successful build #6.
I also added autocommand events after a build fails/succeeds. However, I'm not sure I did it the right way (group user)

### CODE REVIEW

**Comment by cdelledonne** [comment]
Hey, thanks for picking this up. I have a couple of comments:

1. I would prefer to introduce one feature at a time. In this case, either an option to close the CMake window after successful build, or User autocommands. I would prefer the latter, as it provides more flexibility, and can be used to close the window as well as any other thing (as you pointed out in #7).

2. For User autocommands, I would prefer to have a new section in `doc/cmake.txt`, e.g. between "Mappings" and "Configuration", which we could call "Events". We could also add one or two examples in there. Remember to run `vim -u NONE -c "helptags doc | q"` after adding new tags to the docs.

3. From `:help <nomodeline>`:
    > After applying the autocommands the modelines are processed, so that their settings overrule the settings from autocommands when editing a file. This is skipped if <nomodeline> is specified. You probably want to use <nomodeline> for events not used when loading a buffer, such as User. Modelines are also skipped when no matching autocommands were executed.

    This makes me think that we should to `doautocmd <nomodeline> User ...`

4. The events `CMakeBuildFailed` and `CMakeBuildSucceeded` should only trigger when calling `:CMakeBuild`, and not `:CMakeGenerate`, right?

5. Could you also update the CHANGELOG? I forgot to add this step inside CONTRIBUTING.md, will do that soon.

Looking forward to your updates, and thanks again!

**Comment by AlexAUT** [comment]
Sorry for the late reply, I'm currently moving. I will implement your feedback this weekend.

**Comment by cdelledonne** [comment]
No problem, take your time.

By the way, you might want to rebase onto `master` before proceeding. And regarding the examples in the docs, you could show how to jump to the first error in case of a `CMakeBuildFailed` (e.g. using the quickfix command `:cfirst`).

Thanks again!

**Comment by cdelledonne** [comment]
Hi, do you still plan to work on this? Just checking.

**Comment by AlexAUT** [comment]
Hi, sorry the PR got lost after I was moving. I will adjust it during the next few days.

**Comment by AlexAUT** [comment]
@cdelledonne: I only kept the autocmd and adjusted the docs based on your feedback above

File: doc/cmake.txt [hunk 1]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
 
 <C-C>			Stop the current command.
 
+==============================================================================
+Events							*cmake-events*
```
Comment (cdelledonne) [inline doc/cmake.txt:214]: Please capitalize "EVENTS" for consistency with other sections

File: doc/cmake.txt [hunk 2]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+
+To customize behaviour after a command has finished, Vim-CMake calls user
```
Comment (cdelledonne) [inline doc/cmake.txt:216]: I would slightly edit this into

To customize *the* behaviour after a *`:CMakeBuild`* command has finished, Vim-CMake *defines some* user *events*.

(the words in italics are just to point at the changes)

File: doc/cmake.txt [hunk 3]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+autocommands
+
+List of events~
```
Comment (cdelledonne) [inline doc/cmake.txt:219]: Not really needed

File: doc/cmake.txt [hunk 4]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+
+CMakeBuildFailed
```
Comment (cdelledonne) [inline doc/cmake.txt:221]: Please format like e.g. the Global <Plug> Mappings, i.e. event name on the left, and small description on the right (indented like in the <Plug> Mappings as well)

File: doc/cmake.txt [hunk 5]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+CMakeBuildSucceeded
+
+Example usage of CMakeBuildFailed to jump to the first error
```
Comment (cdelledonne) [inline doc/cmake.txt:224]: Please wrap CMakeBuildFailed in backticks

File: doc/cmake.txt [hunk 6]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+>
+      let g:jump_on_error = 0 " We do not want to focus the console
```
Comment (cdelledonne) [inline doc/cmake.txt:226]: Indent all lines with a single tab character please

File: doc/cmake.txt [hunk 7]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+      augroup vim-cmake-group
+	autocmd User CMakeBuildFailed :cfirst
+      augroup END
+<
+
```
Comment (cdelledonne) [inline doc/cmake.txt:231]: Please remove this empty line

File: doc/cmake.txt [hunk 8]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+Example usage of CMakeBuildSucceeded to close the Vim-CMake console
```
Comment (cdelledonne) [inline doc/cmake.txt:232]: Please wrap CMakeBuildSucceeded in backticks

File: doc/cmake.txt [hunk 9]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+>
+      augroup vim-cmake-group
+	autocmd User CMakeBuildSucceeded call cmake#console#Close()
+      augroup END
+<
+
```
Comment (cdelledonne) [inline doc/cmake.txt:238]: Please remove this empty line

File: doc/cmake.txt [hunk 10]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+>
+      let g:jump_on_error = 0 " We do not want to focus the console
+      augroup vim-cmake-group
+	autocmd User CMakeBuildFailed :cfirst
+      augroup END
+<
+
+Example usage of CMakeBuildSucceeded to close the Vim-CMake console
+>
+      augroup vim-cmake-group
```
Comment (cdelledonne) [inline doc/cmake.txt:234]: Indent all lines with a single tab character please

File: doc/cmake.txt [hunk 11]
```diff
@@ -209,6 +210,32 @@ cq			Close the CMake console window.
...
+	autocmd User CMakeBuildSucceeded call cmake#console#Close()
```
Comment (cdelledonne) [inline doc/cmake.txt:235]: cmake#console#Close() is not part of the user Vim-CMake API, please replace "cmake#console#Close()" with "CMakeClose"

File: doc/cmake.txt [hunk 12]
```diff
@@ -272,6 +299,17 @@ g:cmake_root_markers (default: `['.git', '.svn']`)
 			project root.  When Vim-CMake is loaded, it looks for
 			the project root starting from the CWD.
 
+Vim-CMake provides a set of custom events to trigger further actions. For
```
Comment (cdelledonne) [inline doc/cmake.txt:302]: This paragraph is redundant here, as we already have an EVENTS section

File: README.md [hunk 13]
```diff
@@ -126,6 +126,14 @@ follows.
 | `g:cmake_link_compile_commands` | `0`                |
 | `g:cmake_root_markers`          | `['.git', '.svn']` |
 
+Vim-CMake provides a set of custom events to trigger further actions.
```
Comment (cdelledonne) [inline README.md:129]: Could we move this paragraph just above the "### Quickfix list" section, and make it into an "### Events" section, to match the order in doc/cmake.txt?

File: README.md [hunk 14]
```diff
@@ -126,6 +126,14 @@ follows.
...
+Run `:help cmake` for an extensive documentation of all configuration options and examples
+
+| Event                           | Description                              |
+|:--------------------------------|:-----------------------------------------|
+| `User CMakeBuildSucceeded`      | `Triggered on a successful ':CMakeBuild'`|
+| `User CMakeBuildFailed`         | `Triggered after a failed ':CMakeBuild'` |
```
Comment (cdelledonne) [inline README.md:135]: Only wrap command names in backticks, not the "Triggered on a successful" part

**Review by cdelledonne — CHANGES_REQUESTED** [review]
Thanks a lot for picking this up again. I have a few more style comments to be addressed, sorry for being picky on that.

File: doc/cmake.txt (hunk 6 above)
Comment (cdelledonne) [reply doc/cmake.txt:226]: Or I actually mean start all lines with one tab character, and add e.g. two or four spaces (or another tab character) for indented lines

**Comment by AlexAUT** [comment]
Applied the requested changes :)

**Review by cdelledonne — APPROVED** [review]

**Comment by cdelledonne** [comment]
Thanks a lot! :) 

For next time, there's no need to force-push your branch, we can always squash commits when merging. Actually separating commits might help pin-pointing the changes that were made after a review.

**Pull request merged**